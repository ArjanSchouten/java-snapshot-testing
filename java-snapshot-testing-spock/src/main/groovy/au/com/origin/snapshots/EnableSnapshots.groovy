package au.com.origin.snapshots

import java.lang.annotation.*
import org.spockframework.runtime.extension.ExtensionAnnotation

@Retention(RetentionPolicy.RUNTIME)
@Target([ElementType.TYPE, ElementType.METHOD])
@ExtensionAnnotation(SnapshotExtension)
@interface EnableSnapshots {}
