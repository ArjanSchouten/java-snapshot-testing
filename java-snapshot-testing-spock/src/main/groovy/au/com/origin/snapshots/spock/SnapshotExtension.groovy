package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotConfig
import au.com.origin.snapshots.SnapshotConfigInjector
import au.com.origin.snapshots.SnapshotMatcher
import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo

class SnapshotExtension extends AbstractAnnotationDrivenExtension<EnableSnapshots> implements SnapshotConfigInjector {

    void visitSpecAnnotation(EnableSnapshots annotation, SpecInfo spec) {
        SnapshotMatcher.start(new SpockConfig(), spec.reflection)
    }

    void visitSpec(SpecInfo spec) {
        spec.addListener(new SnapshotSpecListener())
    }

    @Override
    SnapshotConfig getSnapshotConfig() {
        return new SpockConfig()
    }
}