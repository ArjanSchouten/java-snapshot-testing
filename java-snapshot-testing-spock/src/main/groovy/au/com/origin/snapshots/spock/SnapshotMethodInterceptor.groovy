package au.com.origin.snapshots.spock

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.SnapshotVerifier
import org.spockframework.runtime.extension.IMethodInterceptor
import org.spockframework.runtime.extension.IMethodInvocation

// Based on this issue: https://github.com/spockframework/spock/issues/652
class SnapshotMethodInterceptor implements IMethodInterceptor {

    private final SnapshotVerifier snapshotVerifier;

    SnapshotMethodInterceptor(SnapshotVerifier snapshotVerifier) {
        this.snapshotVerifier = snapshotVerifier
    }

    @Override
    void intercept(IMethodInvocation invocation) throws Throwable {
        def parameterCount = invocation.method.reflection.parameterCount
        if (parameterCount > invocation.arguments.length) {
            def newArguments = new Object[parameterCount]
            System.arraycopy invocation.arguments, 0, newArguments, 0, invocation.arguments.length
            invocation.arguments = newArguments
        }
        invocation.method.reflection.parameterTypes.eachWithIndex { type, i ->
            if (Expect.class == type) {
                invocation.arguments[i] = new Expect(snapshotVerifier, invocation.feature.featureMethod.reflection)
            }
        }
        invocation.proceed()
    }
}
