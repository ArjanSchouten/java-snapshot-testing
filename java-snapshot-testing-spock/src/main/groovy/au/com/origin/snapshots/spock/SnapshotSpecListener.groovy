package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotMatcher
import org.spockframework.runtime.AbstractRunListener;
import org.spockframework.runtime.model.*;

class SnapshotSpecListener extends AbstractRunListener {

    void beforeFeature(FeatureInfo feature) {
        SnapshotMatcher.testMethod = feature.featureMethod.reflection
    }

}
