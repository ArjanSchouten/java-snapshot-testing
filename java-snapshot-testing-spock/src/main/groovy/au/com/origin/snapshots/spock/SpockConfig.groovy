package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotConfig

import java.lang.reflect.Method

class SpockConfig implements SnapshotConfig {

    @Override
    String getOutputDir() {
        return "src/test/groovy/"
    }

    @Override
    Class<?> getTestClass() {
        throw new RuntimeException("You forgot to annotate your spec with @EnableSnapshots");
    }

    @Override
    Method getTestMethod(Class<?> testClass) {
        throw new RuntimeException("You forgot to annotate your spec with @EnableSnapshots");
    }
}