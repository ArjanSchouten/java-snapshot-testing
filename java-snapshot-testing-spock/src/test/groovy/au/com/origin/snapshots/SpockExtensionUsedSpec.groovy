package au.com.origin.snapshots

import au.com.origin.snapshots.spock.EnableSnapshots
import spock.lang.Specification

@EnableSnapshots
class SpockExtensionUsedSpec extends Specification {

    def "Should use extension"() {
        when:
        SnapshotMatcher.expect("Hello World").toMatchSnapshot()

        then:
        true
    }

    def "Should use extension again"() {
        when:
        SnapshotMatcher.expect("Hello World", "Hello World Again").toMatchSnapshot()

        then:
        true
    }

}
