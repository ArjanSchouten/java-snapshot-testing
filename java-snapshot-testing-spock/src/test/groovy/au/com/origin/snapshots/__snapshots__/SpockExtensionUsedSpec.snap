au.com.origin.snapshots.SpockExtensionUsedSpec.$spock_feature_0_0=[
Hello World
]


au.com.origin.snapshots.SpockExtensionUsedSpec.$spock_feature_0_1=[
Hello World
Hello World Again
]